contents/projectManagement/testDrivenDevelopment.md

```diff
@@ -0,0 +1,199 @@
+
+# Test-Driven Development
+
+Author: [Yash Chowdhary](https://github.com/yash-chowdhary)
+
+## What is Test-driven Development? 
+
+Typically when you write unit tests, you test out the functionality of code you've already written. Test-driven development (a.k.a TDD) is a software development process that emphasizes the opposite - writing tests before production code. It is a design technique aimed at understanding requirements before writing functional code.
+
+### Development Cycle
+
+Robert Martin describes the "Three Laws of TDD" as - 
+> "First Law: You may not write production code until you have written a failing unit test. 
+>
+> Second Law: You may not write more of a unit test than is sufficient to fail, and not compiling is failing. 
+>
+> Third Law: You may not write more production code than is sufficient to pass the currently failing test.”
+
+Kent Beck, creator of Xtreme Programming and advocate of TDD, states that the general development cycle is as follows[^1]:
+
+1. **Add a test**
+
+    In this phase of the cycle, all you need to focus on is writing tests. Understanding the product's/feature's requirements and specifications is crucial at this stage of the development cycle. 
+
+1. **Run the test**
+
+    In this phase, you check if the new test added fails. This is an important step, as it shows that the new test does not pass without requiring new code because the required behavior already exists, and also rules out false positives (i.e. the test passes when it shouldn't). Furthermore, the reason for its failure must be clear. 
+
+1. **Write functional code**
+
+    Here is when you write code to pass the test. The sole aim at this point is to pass the test and to write code whose functionality is restricted to the scope of the test. It is possible that the code may not be in its best form (style- and syntax-wise), but that's something you need not worry about at this stage.
+
+1. **Re-run test(s)**
+
+    Re-running the test suite and ensuring the test passes not only helps in keeping the cycle running, but also in increasing your confidence in the code that you have just written. This is because it shows that the new code meets the test requirements, and does not break any existing features/requirements. 
+
+1. **Refactor the code**
+
+    As and when you write more production code and test to make sure it meets all your business, product, and/or feature requirements, it's important that the code itself meets a certain standard. Improving code style and readability, employing design patterns, and reducing duplication of code are some of the [many things](https://refactoring.com/catalog/) you can work on.
+
+
+Steps 1-5 outline one **iteration** of the TDD cycle. In practice, several such iterations are required - sometimes even to get the same test or portion of the test to pass.
+
+Figure 1 below shows the general TDD cycle.
+
+<center>
+<pic src="images/tddSteps.jpg" alt="tddSteps" height="420">
```

Comment:
Personally think its better to move this to the top, readers may want to see the overall picture first before delving into the details

Reply:
Addressed. It seems like a good idea